sfe/sfe.go

```diff
@@ -0,0 +1,317 @@
+package sfe
+
+import (
+	"embed"
+	"errors"
+	"fmt"
+	"io/fs"
+	"net/http"
+	"strconv"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/go-jose/go-jose/v4"
+	"github.com/go-jose/go-jose/v4/jwt"
+	"github.com/jmhodges/clock"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+
+	"github.com/letsencrypt/boulder/core"
+
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/metrics/measured_http"
+	rapb "github.com/letsencrypt/boulder/ra/proto"
+	sapb "github.com/letsencrypt/boulder/sa/proto"
+)
+
+const (
+	// The API version should be checked when parsing parameters to quickly deny
+	// a client request. Can be used to mass-invalidate URLs. Must be
+	// concatenated with other path slugs.
+	unpauseAPISlug = "/sfe/v1"
+)
+
+var (
+	unpauseGetForm    = unpauseAPISlug + "/unpause"
+	unpausePostForm   = unpauseAPISlug + "/do-unpause"
+	unpauseSuccessful = unpauseAPISlug + "/unpause-successful"
+
+	//go:embed all:static
+	staticFS embed.FS
+
+	//go:embed all:templates all:pages
+	dynamicFS embed.FS
+)
+
+// SelfServiceFrontEndImpl provides all the logic for Boulder's selfservice
+// frontend web-facing interface, i.e., a portal where a subscriber can unpause
+// their account. Its methods are primarily handlers for HTTPS requests for the
+// various non-ACME functions.
+type SelfServiceFrontEndImpl struct {
+	ra rapb.RegistrationAuthorityClient
+	sa sapb.StorageAuthorityReadOnlyClient
+
+	log blog.Logger
+	clk clock.Clock
+
+	// requestTimeout is the per-request overall timeout.
+	requestTimeout time.Duration
+
+	// unpauseHMACHash is used to validate incoming JWT signatures on the
+	// unpause endpoint and should be shared by the SFE and WFE.
+	unpauseHMACHash []byte
```

Comment:
Based on the commented-out code in sfe/main.go, I believe this (and all other instances of `unpauseHMACHash`) should be renamed to `unpauseHMACKey`.